libsrc/common/getcpu.s

```diff
@@ -26,52 +27,119 @@
 .p816                           ; Enable 65816 instructions
 
 _getcpu:
+
         lda     #0
         inc     a               ; .byte $1A ; nop on nmos, inc on every cmos
         cmp     #1
-        bcc     @L8
+        bcc     @IsNMOS
 
 ; This is at least a 65C02, check for a 65CE02/4510
 
         .byte   $42,$EA         ; neg on 65CE02/4510, nop #$EA on 65C02, wdm $EA on 65816
         cmp     #1
-        beq     @L6
+        beq     @HasINCA
 
 ; This is at least a 65CE02, check for 4510
 
         lda     #5              ; CPU_65CE02 constant
+        ldx     #0              ; to make sure MAP doesn't do anything, the upper nybl of X and Z must be clear
         .byte   $5C             ; map on 4510, aug on 65CE02 (acts like 4 byte nop)
         lda     #3              ; CPU_4510 constant
         nop
-        bne     @L9
+        cmp     #5
+        beq     @LoadXAndReturn
+
+; It is either a 4510 (C65) or a 45GS02 (MEGA65)
+
+        ; 45GS02 supports 32-bit ZP indirect, so use that to check CPU type
+        ; without requiring a functioning MEGA65 hypervisor.
+        ; We setup a read of $200FF, then store a different value in $00FF
+        ; and then re-read $200FF to see if it is unchanged.
+
+        ; Save the 32-bit ZP pointer and data byte
+        ldx     #4
+@L10:   lda     $fb,x
+        sta     @GetCPUTemp,x
```

Comment:
You don't need to "steal" any zero-page locations.  The run-time code uses many locations that are available.  Most of them aren't being used while `getcpu()` is running; and, their contents don't need to be saved.

Look at [zeropage.s](https://github.com/cc65/cc65/blob/master/libsrc/runtime/zeropage.s).  You must `.include "zeropage.inc"`.  Then, you can use any group of four bytes except `sp:` and `regbank:`.

Reply:
There's even documentation: https://github.com/cc65/wiki/wiki/Using-runtime-zeropage-locations-in-assembly-language

Reply:
Now uses zeropage.inc to use existing ZP storage.